src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/port_proxy.cc

```diff
@@ -0,0 +1,175 @@
+  recv_port_->Join();
+}
+py::object RecvPortProxy::Peek() {
+  MetaDataPtr metadata = recv_port_->Peek();
+  return MDataToObject_(metadata);
+}
+std::string RecvPortProxy::Name() {
+  return recv_port_->Name();
+}
+size_t RecvPortProxy::Size() {
+  return recv_port_->Size();
+}
+
+int trick() {
+    // to solve the warning "converting to non-pointer type 'int'
+    // from NULL [-Wconversion-null] import_array()"
+    _import_array();
+    return 0;
+}
+
+const int tricky_var = trick();
+
+MetaDataPtr SendPortProxy::MDataFromObject_(py::object* object) {
+  PyObject *obj = object->ptr();
+  LAVA_LOG(LOG_LAYER, "start MDataFromObject\n");
+  if (!PyArray_Check(obj)) {
+    LAVA_LOG_ERR("The Object is not array tp is %s\n", Py_TYPE(obj)->tp_name);
+    exit(-1);
+  }
+
+  LAVA_LOG(LOG_LAYER, "check obj achieved\n");
+
+  auto array = reinterpret_cast<PyArrayObject*> (obj);
+  if (!PyArray_ISWRITEABLE(array)) {
+    LAVA_LOG(LOG_LAYER, "The array is not writeable\n");
+  }
+
+  // var from numpy
+  int32_t ndim = PyArray_NDIM(array);
+  auto dims = PyArray_DIMS(array);
+  auto strides = PyArray_STRIDES(array);
+  void* data_ptr = PyArray_DATA(array);
+  // auto dtype = PyArray_Type(array);  // no work
+  auto dtype = array->descr->type_num;
+  auto element_size_in_bytes = PyArray_ITEMSIZE(array);
+  auto tsize = PyArray_SIZE(array);
+
+  // set metadata
+  MetaDataPtr metadata = std::make_shared<MetaData>();
+  metadata->nd = ndim;
+  for (int i = 0; i < ndim; i++) {
+    metadata->dims[i] = dims[i];
+    metadata->strides[i] = strides[i]/element_size_in_bytes;
+    if (strides[i] % element_size_in_bytes != 0) {
+      LAVA_LOG_ERR("numpy array stride not a multiple of element bytes\n");
+    }
+  }
+  metadata->type = dtype;
+  metadata->mdata = data_ptr;
+  metadata->elsize = element_size_in_bytes;
+  metadata->total_size = tsize;
+
+  return metadata;
+}
+
+py::object RecvPortProxy::MDataToObject_(MetaDataPtr metadata) {
+  if (metadata == nullptr)
+    return py::cast(0);
+
+  std::vector<npy_intp> dims(metadata->nd);
+  std::vector<npy_intp> strides(metadata->nd);
+
+  for (int i = 0; i < metadata->nd; i++) {
+    dims[i] = metadata->dims[i];
+    strides[i] = metadata->strides[i] * metadata->elsize;
+  }
+
+  PyObject *array = PyArray_New(
+    &PyArray_Type,
+    metadata->nd,
+    dims.data(),
+    metadata->type,
+    strides.data(),
+    metadata->mdata,
+    metadata->elsize,
+    NPY_ARRAY_ALIGNED | NPY_ARRAY_WRITEABLE,
+    nullptr);
+
+  if (!array)
+    return py::cast(0);
+  LAVA_DEBUG(LOG_LAYER, "Set PyObject capsule, mdata: %p\n", metadata->mdata);
+  PyObject *capsule = PyCapsule_New(metadata->mdata, nullptr,
+                                      [](PyObject *capsule) {
+    void *memory = PyCapsule_GetPointer(capsule, nullptr);
+    LAVA_DEBUG(LOG_LAYER, "PyObject cleaned, free memory: %p, size: %lu.\n",
+                memory, *(reinterpret_cast<uint64_t*>(memory) - 1)&(~0x7));
+    free(memory);});
```

Comment:
I am very warry of all these free()s. It takes too much effort to think about who needs to free when and when an exception might be thrown.

I am pretty sure we can use RAIImdata (and other malloc'd data).

Could we:
1) Make mdata a std::vector<char>.
1a) Or make mdata a std::unique_ptr<char[]>; which can be allocated with dynamic size via std::make_unique<char[]>(size) since C++14.
2) Then, when calling PyCapsule_New:
2a) Pass in (new std::vector<char>(std::move(metadata->mdata))) as the first argument, or:
2b) Pass in (new std::unique_ptr<char[]>(std::move(metadata->mdata))) as the first argument.
3) In the PyCapsule_New's lambda, use the following as the first line:
3a) std::unique_ptr<std::vector<char>> memory(PyCapsule_GetPointer(capsule, nullptr)); or:
3b) std::unique_ptr<std::unique_ptr<char[]>> memory(PyCapsule_GetPointer(capsule, nullptr));

That way, even if LAVA_DEBUG throws, you will still free the memory. Then you could delete all the calls to free() and not have to worry about it.

Reply:
Thanks for u suggestion.
We hope to use STL container or smart pointer to simplify free/array operations, but the input data type of [`PyArray_New`](https://numpy.org/doc/1.13/reference/c-api.array.html) is `void *` as following:
```cpp
PyObject* PyArray_New(PyTypeObject* subtype, int nd, npy_intp* dims, int type_num, npy_intp* strides, void* data, int itemsize, int flags, PyObject* obj
```
so cannot adopt u suggestion.

Reply:
What kinds of errors show up if an STL container or smart pointer is used?

There is a performance penalty if we use a smart pointer, since we'll need to allocate an extra smart pointer to pass ownership through PyCapsule_GetPointer. Despite the penalty, I think it's worth figuring out. It sounds like you hope there's a way too, since it'll be easier to avoid memory leaks.

There's got to be a way to make this work without calling malloc/calloc directly. In the worst case, we should be able to create a custom allocator that calls malloc/calloc.

PyArray_New takes a void* for data. void* on it's own can be aligned to any char boundary, so it is legal to static_cast<void*> a char*. Is the error that python or numpy require more strict alignment for SIMD instructions or something like that? 

The following makes it seem like this should be possible in C++17 without a custom allocator: https://stackoverflow.com/questions/8456236/how-is-a-vectors-data-aligned 

Following hints from that page, would something like the following work:

auto kPyArrayAlignment = std::align_val_t{64};
using SmartBufferPtr = std::unique_ptr<char[]>;

SmartBufferPtr mdata;
metadata->mdata = SmartBufferPtr(new (kPyArrayAlignment) char[size]);

PyArray_New(..., static_cast<void*>(metadata->mdata->get()), ...);

SmartBufferPtr *mdata_py = new SmartBufferPtr(std::move(metadata->mdata));
PyCapsule_New(mdata_py, nullptr, [](PyObject *capsule) {
    std::unique_ptr<SmartBufferPtr> memory(static_cast<SmartBufferPtr*>(PyCapsule_GetPointer(capsule, nullptr)));
}

If that doesn't work, would a custom allocator work?

Reply:
Maybe it works. i have one question:
- Why not use malloc/calloc? <br>
We have validated current implement with all unit tests. There aren't the issue of free mdata.

During 2nd release, we don't have enough effort to evaluate your suggestion.
I think we could discuss this comment after 2nd release.